modules/eloqua_api_auth_fallback/src/Commands/EloquaAuthTokensGenerate.php

```diff
@@ -0,0 +1,104 @@
+
+  /**
+   * Immutable Config.
+   *
+   * @var \Drupal\Core\Config\Config|\Drupal\Core\Config\ImmutableConfig
+   */
+  protected $configSettings;
+
+  /**
+   * The logger factory.
+   *
+   * @var \Drupal\Core\Logger\LoggerChannelFactoryInterface
+   */
+  protected $loggerFactory;
+
+  /**
+   * Eloqua API client service.
+   *
+   * @var \Drupal\eloqua_api_redux\Service\EloquaApiClient
+   */
+  protected $eloquaApiClient;
+
+  /**
+   * EloquaAuthTokensGenerate constructor.
+   *
+   * @param \Drupal\Core\Config\ConfigFactory $config_factory
+   *   An instance of ConfigFactory.
+   * @param \Drupal\Core\Logger\LoggerChannelFactoryInterface $logger_factory
+   *   LoggerChannelFactoryInterface.
+   * @param \Drupal\eloqua_api_redux\Service\EloquaApiClient $eloqua_api_client
+   *   Eloqua API client service.
+   */
+  public function __construct(ConfigFactory $config_factory,
+                              LoggerChannelFactoryInterface $logger_factory,
+                              EloquaApiClient $eloqua_api_client) {
+    parent::__construct();
+    $this->configSettings = $config_factory->get('eloqua_api_auth_fallback.settings');
+    $this->loggerFactory = $logger_factory;
+    $this->eloquaApiClient = $eloqua_api_client;
+  }
+
+  /**
+   * Calls eloqua authentication API service to generate tokens.
+   *
+   * Access and refresh tokens are generated using resource owner password
+   * credentials grant method.
+   *
+   * @command eloqua_api_auth_fallback:generate-tokens
+   * @aliases eloqua-gt
+   * @usage eloqua_api_auth_fallback:generate-tokens
+   *   Calls eloqua authentication API and generate tokens
+   *   using Resource Owner Password Credentials grant.
+   */
+  public function generateTokens() {
+    $sitename = $this->configSettings->get('sitename');
+    $username = $this->configSettings->get('username');
+    $password = $this->configSettings->get('password');
+    if (empty($sitename) || empty($username) || empty($password)) {
+      $message = 'Eloqua authentication credentials are not set.';
+      $this->output()
+        ->writeln($message);
+      $this->loggerFactory->get('eloqua_api_auth_fallback')->error($message);
+      return FALSE;
+    }
+
+    $params = [
+      'grant_type' => 'password',
+      'scope' => 'full',
+      'username' => $sitename . '\\' . $username,
+      'password' => $password,
+    ];
+    $response = $this->eloquaApiClient->doTokenRequest($params);
+
+    if (!empty($response) && !empty($response['refresh_token'])) {
+      $message = 'Eloqua refresh and access tokens updated successfully.';
+      $this->output()
+        ->writeln($message);
+      $this->loggerFactory->get('eloqua_api_auth_fallback')->info($message);
+      return TRUE;
+    }
+    else {
+      $message = 'Error updating refresh and access tokens.';
+      $this->output()
+        ->writeln($message);
+      $this->loggerFactory->get('eloqua_api_auth_fallback')->error($message);
```

Comment:
[EloquaApiClient::doTokenRequest() already has logging on failed request](https://github.com/dakkusingh/eloqua_api_redux/blob/7e0752e7867ccf2062f51de989a5ee803b27bceb/src/Service/EloquaApiClient.php#L202). I dont think log duplication is required here.

Reply:
Removed